File: data/recommend.py
"""
recommend.py - 유사 난이도 패턴 추천

현재 선택된 패턴의 floor 값을 기준으로
±floor_range 범위 내 패턴을 찾고,
RecordDB의 rate를 붙여 정렬해 반환한다.

난이도 체계:
  비공식(floorName 있음): NM/HD/MX/SC 공통 척도 → 전 난이도 대상
  공식(floorName 없음):   NM/HD/MX vs SC 별도 체계
                          → SC 선택 시 SC끼리, NM/HD/MX 선택 시 NM/HD/MX끼리만

정렬 우선순위:
  1. 기록 있음(rate > 0) → rate 낮은 순  (약한 패턴 우선)
  2. 미탐색(None)        → floor 오름차순
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from data.varchive import VArchiveDB, DIFFICULTIES
from data.record_manager import RecordManager
from constants import SC_GROUP, NHM_GROUP, DIFF_COLORS


def _parse_floor_value(floor_name: Optional[str]) -> Optional[float]:
    """'15.2' → 15.2, None → None"""
    if not floor_name:
        return None
    try:
        return float(floor_name)
    except ValueError:
        return None


def _diff_group(difficulty: str) -> str:
    """공식 난이도 체계에서 같은 그룹인지 판별하기 위한 그룹 키."""
    return "SC" if difficulty in SC_GROUP else "NHM"


@dataclass
class RecommendEntry:
    song_id:     int
    song_name:   str
    composer:    str
    button_mode: str
    difficulty:  str
    level:       Optional[int]
    floor:       Optional[float]
    floor_name:  Optional[str]
    rate:        Optional[float]
    is_max_combo: bool = False

    @property
    def has_record(self) -> bool:
        return self.rate is not None

    @property
    def is_played(self) -> bool:
        return self.rate is not None

    @property
    def is_perfect_play(self) -> bool:
        return self.rate is not None and self.rate >= 100.0
        
    @property
    def is_max_combo_play(self) -> bool:
        return self.is_max_combo


@dataclass
class RecommendResult:
    entries:              list[RecommendEntry]
    avg_rate:             float
    has_record_count:     int
    total_count:          int

    @staticmethod
    def empty() -> RecommendResult:
        return RecommendResult([], -1.0, 0, 0)


@dataclass(frozen=True)
class FloorCacheKey:
    button_mode: str
    scale_type: str
    floor: float


@dataclass
class FloorRateSummary:
    total_count: int = 0
    has_record_count: int = 0
    rate_sum: float = 0.0

    @property
    def avg_rate(self) -> float:
        if self.has_record_count <= 0:
            return -1.0
        return self.rate_sum / self.has_record_count


class Recommender:
    def __init__(self, varchive_db: VArchiveDB, record_db: RecordManager):
        self.vdb = varchive_db
        self.rdb = record_db
        self._floor_rate_cache: dict[FloorCacheKey, FloorRateSummary] = {}
        self._floor_rate_dirty: dict[FloorCacheKey, bool] = {}
        self._floor_patterns: dict[FloorCacheKey, list[tuple[int, str, str]]] = {}
        self._record_to_floor_key: dict[tuple[int, str, str], FloorCacheKey] = {}
        self._cache_index_ready: bool = False

    def recommend(
        self,
        song_id: int,
        button_mode: str,
        difficulty: str,
        floor_range: float = 0.0,
        max_results: int = 6,
        same_mode_only: bool = True,
    ) -> RecommendResult:
        """현재 패턴과 floor가 유사한 패턴 목록 반환."""
        current_song = self.vdb.search_by_id(song_id)
        if not current_song:
            return RecommendResult.empty()

        current_pattern = (
            current_song.get("patterns", {})
            .get(button_mode, {})
            .get(difficulty)
        )
        if not current_pattern:
            return RecommendResult.empty()

        floor_name_ref = current_pattern.get("floorName")
        ref_floor      = _parse_floor_value(floor_name_ref)
        use_official   = ref_floor is None

        if use_official:
            ref_floor    = float(current_pattern.get("level", 0))
            ref_diff_grp = _diff_group(difficulty)
        else:
            ref_diff_grp = ""

        # 1. 후보 수집
        candidates = self._get_candidates(
            song_id, button_mode, difficulty,
            ref_floor, use_official, ref_diff_grp,
            floor_range, same_mode_only
        )

        if not candidates:
            return RecommendResult.empty()

        # 2. RecordDB 레이트 병합
        self._merge_record_rates(candidates)

        # 3. 정렬 및 결과 반환
        def sort_key(e: RecommendEntry) -> tuple:
            if e.is_played:
                return (0, e.rate, e.floor or 0.0)
            else:
                return (1, e.floor or 0.0, 0.0)

        candidates.sort(key=sort_key)
        summary = self._get_summary_from_cache(
            song_id=song_id,
            button_mode=button_mode,
            difficulty=difficulty,
            ref_floor=ref_floor,
            use_official=use_official,
            floor_range=floor_range,
            same_mode_only=same_mode_only,
        )
        return RecommendResult(
            candidates[:max_results],
            summary.avg_rate,
            summary.has_record_count,
            summary.total_count,
        )

    def _get_candidates(
        self,
        target_song_id: int,
        target_mode: str,
        target_diff: str,
        ref_floor: float,
        use_official: bool,
        ref_diff_grp: str,
        floor_range: float,
        same_mode_only: bool
    ) -> list[RecommendEntry]:
        """주어진 조건에 맞는 추천 후보를 수집한다."""
        from data.varchive import BUTTON_MODES
        modes_to_check = [target_mode] if same_mode_only else BUTTON_MODES
        candidates = []

        for song in self.vdb.songs:
            try:
                sid = int(song.get("title", 0))
            except (ValueError, TypeError):
                continue
            patterns = song.get("patterns", {})

            for mode in modes_to_check:
                mode_patterns = patterns.get(mode, {})
                for diff in DIFFICULTIES:
                    p = mode_patterns.get(diff)
                    if not p:
                        continue

                    cand_floor_name = p.get("floorName")
                    cand_floor = _parse_floor_value(cand_floor_name)

                    if use_official:
                        if cand_floor is not None:
                            continue
                        if _diff_group(diff) != ref_diff_grp:
                            continue
                        cand_floor = float(p.get("level", 0))
                    else:
                        if cand_floor is None:
                            continue

                    if abs(cand_floor - ref_floor) > floor_range:
                        continue

                    if sid == target_song_id and mode == target_mode and diff == target_diff:
                        continue

                    candidates.append(RecommendEntry(
                        song_id=sid,
                        song_name=song.get("name", ""),
                        composer=song.get("composer", ""),
                        button_mode=mode,
                        difficulty=diff,
                        level=p.get("level"),
                        floor=cand_floor,
                        floor_name=cand_floor_name,
                        rate=None,
                    ))
        return candidates

    def _merge_record_rates(self, candidates: list[RecommendEntry]):
        """후보 목록에 RecordDB의 레이트 정보를 병합한다."""
        if not self.rdb.is_ready:
            return

        all_ids = list({c.song_id for c in candidates})
        rate_map = self.rdb.get_rate_map(all_ids)

        for entry in candidates:
            key = (entry.song_id, entry.button_mode, entry.difficulty)
            if key in rate_map:
                rec = rate_map[key]
                entry.rate = rec["rate"]
                entry.is_max_combo = rec["is_max_combo"]

    def _ensure_floor_rate_cache(self):
        if not self._cache_index_ready:
            self._build_floor_cache_index()

        full_dirty, dirty_keys = self.rdb.consume_dirty_info()
        if full_dirty:
            for key in self._floor_patterns:
                self._floor_rate_dirty[key] = True
        else:
            for record_key in dirty_keys:
                floor_key = self._record_to_floor_key.get(record_key)
                if floor_key is not None:
                    self._floor_rate_dirty[floor_key] = True

        dirty_floor_keys = [k for k, is_dirty in self._floor_rate_dirty.items() if is_dirty]
        if not dirty_floor_keys:
            return

        all_song_ids = []
        for song in self.vdb.songs:
            try:
                song_id = int(song.get("title", 0))
            except (ValueError, TypeError):
                continue
            all_song_ids.append(song_id)
        rate_map = self.rdb.get_rate_map(list(set(all_song_ids))) if all_song_ids else {}
        for key in dirty_floor_keys:
            entries = self._floor_patterns.get(key, [])
            summary = FloorRateSummary(total_count=len(entries))
            for song_id, mode, diff in entries:
                rec = rate_map.get((song_id, mode, diff))
                if not rec:
                    continue
                rate = rec.get("rate", 0.0)
                if rate <= 0.0:
                    continue
                summary.has_record_count += 1
                summary.rate_sum += rate
            self._floor_rate_cache[key] = summary
            self._floor_rate_dirty[key] = False

    def _build_floor_cache_index(self):
        floor_patterns: dict[FloorCacheKey, list[tuple[int, str, str]]] = {}
        record_to_floor_key: dict[tuple[int, str, str], FloorCacheKey] = {}

        for song in self.vdb.songs:
            try:
                song_id = int(song.get("title", 0))
            except (ValueError, TypeError):
                continue
            patterns = song.get("patterns", {})
            for mode, mode_patterns in patterns.items():
                for diff in DIFFICULTIES:
                    p = mode_patterns.get(diff)
                    if not p:
                        continue
                    floor_name = p.get("floorName")
                    floor_val = _parse_floor_value(floor_name)
                    if floor_val is None:
                        level = p.get("level")
                        if level is None:
                            continue
                        floor_val = float(level)
                        scale_type = "OFFICIAL_SC" if diff in SC_GROUP else "OFFICIAL_NHM"
                    else:
                        scale_type = "UNOFFICIAL"
                    key = FloorCacheKey(mode, scale_type, floor_val)
                    record_key = (song_id, mode, diff)
                    floor_patterns.setdefault(key, []).append(record_key)
                    record_to_floor_key[record_key] = key

        self._floor_patterns = floor_patterns
        self._record_to_floor_key = record_to_floor_key
        self._floor_rate_cache = {
            key: FloorRateSummary(total_count=len(entries))
            for key, entries in self._floor_patterns.items()
        }
        self._floor_rate_dirty = {key: True for key in self._floor_patterns}
        self._cache_index_ready = True

    def _get_summary_from_cache(
        self,
        song_id: int,
        button_mode: str,
        difficulty: str,
        ref_floor: float,
        use_official: bool,
        floor_range: float,
        same_mode_only: bool,
    ) -> FloorRateSummary:
        self._ensure_floor_rate_cache()

        if use_official:
            scale_type = "OFFICIAL_SC" if difficulty in SC_GROUP else "OFFICIAL_NHM"
        else:
            scale_type = "UNOFFICIAL"

        from data.varchive import BUTTON_MODES
        modes = [button_mode] if same_mode_only else BUTTON_MODES

        total = 0
        has_record = 0
        rate_sum = 0.0
        for key, summary in self._floor_rate_cache.items():
            if key.button_mode not in modes:
                continue
            if key.scale_type != scale_type:
                continue
            if abs(key.floor - ref_floor) > floor_range:
                continue
            total += summary.total_count
            has_record += summary.has_record_count
            rate_sum += summary.rate_sum

        return FloorRateSummary(
            total_count=max(0, total),
            has_record_count=max(0, has_record),
            rate_sum=max(0.0, rate_sum),
        )

File: rust/overmax_data/src/service/recommend.rs
use crate::community::client::VArchiveDB;
use crate::service::record_manager::{RecordManager, RecordSource};
use std::cmp::Ordering;
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering as AtomicOrdering};
use std::sync::{Arc, Mutex};

const DIFFICULTIES: &[&str] = &["NM", "HD", "MX", "SC"];
const SC_GROUP: &[&str] = &["SC"];
const MODES: &[&str] = &["4B", "5B", "6B", "8B"];

type RecordKey = (i32, String, String);

#[derive(Debug, Clone, PartialEq)]
pub struct RecommendEntry {
    pub song_id: i32,
    pub song_name: String,
    pub composer: String,
    pub button_mode: String,
    pub difficulty: String,
    pub level: Option<u32>,
    pub floor: Option<f64>,
    pub floor_name: Option<String>,
    pub rate: Option<f64>,
    pub is_max_combo: bool,
}

impl RecommendEntry {
    pub fn is_played(&self) -> bool {
        self.rate.is_some()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct RecommendResult {
    pub entries: Vec<RecommendEntry>,
    pub avg_rate: f64,
    pub has_record_count: usize,
    pub total_count: usize,
}

impl RecommendResult {
    pub fn empty() -> Self {
        Self {
            entries: Vec::new(),
            avg_rate: -1.0,
            has_record_count: 0,
            total_count: 0,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct FloorCacheKey {
    pub button_mode: String,
    pub scale_type: String,
    pub floor_millis: i64,
}

#[derive(Debug, Clone)]
pub struct FloorRateSummary {
    pub total_count: usize,
    pub has_record_count: usize,
    pub rate_sum: f64,
}

impl FloorRateSummary {
    pub fn new(total_count: usize) -> Self {
        Self {
            total_count,
            has_record_count: 0,
            rate_sum: 0.0,
        }
    }

    pub fn avg_rate(&self) -> f64 {
        if self.has_record_count == 0 {
            return -1.0;
        }
        self.rate_sum / self.has_record_count as f64
    }
}

pub struct Recommender {
    vdb: Arc<VArchiveDB>,
    rdb: Arc<RecordManager>,
    floor_rate_cache: Mutex<HashMap<FloorCacheKey, FloorRateSummary>>,
    floor_rate_dirty: Mutex<HashMap<FloorCacheKey, bool>>,
    floor_patterns: Mutex<HashMap<FloorCacheKey, Vec<RecordKey>>>,
    record_to_floor_key: Mutex<HashMap<RecordKey, FloorCacheKey>>,
    cache_index_ready: AtomicBool,
}

struct CandidateSearchParams<'a> {
    target_song_id: i32,
    target_mode: &'a str,
    target_diff: &'a str,
    ref_floor: f64,
    use_official: bool,
    ref_diff_grp: &'a str,
    floor_range: f64,
    same_mode_only: bool,
}

impl Recommender {
    pub fn new(vdb: Arc<VArchiveDB>, rdb: Arc<RecordManager>) -> Self {
        Self {
            vdb,
            rdb,
            floor_rate_cache: Mutex::new(HashMap::new()),
            floor_rate_dirty: Mutex::new(HashMap::new()),
            floor_patterns: Mutex::new(HashMap::new()),
            record_to_floor_key: Mutex::new(HashMap::new()),
            cache_index_ready: AtomicBool::new(false),
        }
    }

    fn parse_floor_value(floor_name: Option<&String>) -> Option<f64> {
        floor_name.and_then(|s| s.parse::<f64>().ok())
    }

    fn floor_to_millis(floor: f64) -> i64 {
        (floor * 1000.0).round() as i64
    }

    fn diff_group(diff: &str) -> &'static str {
        if SC_GROUP.contains(&diff) {
            "SC"
        } else {
            "NHM"
        }
    }

    pub fn recommend(
        &self,
        song_id: i32,
        button_mode: &str,
        difficulty: &str,
        floor_range: f64,
        max_results: usize,
        same_mode_only: bool,
    ) -> RecommendResult {
        let current_song = match self.vdb.search_by_id(song_id) {
            Some(s) => s,
            None => return RecommendResult::empty(),
        };

        let current_pattern = crate::community::client::Mode::from_str(button_mode)
            .zip(crate::community::client::Difficulty::from_str(difficulty))
            .and_then(|(m, d)| current_song.patterns[m as usize][d as usize].as_ref());

        let p = match current_pattern {
            Some(p) => p,
            None => return RecommendResult::empty(),
        };

        let ref_floor = Self::parse_floor_value(p.floor_name.as_ref());
        let use_official = ref_floor.is_none();

        let (final_ref_floor, ref_diff_grp) = if let Some(floor) = ref_floor {
            (floor, "")
        } else {
            (p.level.unwrap_or(0) as f64, Self::diff_group(difficulty))
        };

        let mut candidates = self.get_candidates(CandidateSearchParams {
            target_song_id: song_id,
            target_mode: button_mode,
            target_diff: difficulty,
            ref_floor: final_ref_floor,
            use_official,
            ref_diff_grp,
            floor_range,
            same_mode_only,
        });

        self.merge_record_rates(&mut candidates);

        candidates.sort_by(|a, b| {
            if a.is_played() && !b.is_played() {
                Ordering::Less
            } else if !a.is_played() && b.is_played() {
                Ordering::Greater
            } else if let (Some(ra), Some(rb)) = (a.rate, b.rate) {
                ra.partial_cmp(&rb)
                    .unwrap_or(Ordering::Equal)
                    .then_with(|| {
                        a.floor
                            .unwrap_or(0.0)
                            .partial_cmp(&b.floor.unwrap_or(0.0))
                            .unwrap_or(Ordering::Equal)
                    })
            } else {
                a.floor
                    .unwrap_or(0.0)
                    .partial_cmp(&b.floor.unwrap_or(0.0))
                    .unwrap_or(Ordering::Equal)
            }
        });

        let summary = self.get_summary_from_cache(
            button_mode,
            difficulty,
            final_ref_floor,
            use_official,
            floor_range,
            same_mode_only,
        );

        candidates.truncate(max_results);

        RecommendResult {
            entries: candidates,
            avg_rate: summary.avg_rate(),
            has_record_count: summary.has_record_count,
            total_count: summary.total_count,
        }
    }

    fn get_candidates(&self, params: CandidateSearchParams) -> Vec<RecommendEntry> {
        let modes_to_check = if params.same_mode_only {
            vec![params.target_mode]
        } else {
            MODES.to_vec()
        };

        let mut candidates = Vec::new();

        for song in &self.vdb.songs {
            let sid = match song.title.parse::<i32>() {
                Ok(id) => id,
                Err(_) => continue,
            };

            for mode in &modes_to_check {
                if let (Some(m), Some(d_list)) = (
                    crate::community::client::Mode::from_str(mode),
                    Some(DIFFICULTIES),
                ) {
                    for diff in d_list {
                        if let Some(d) = crate::community::client::Difficulty::from_str(diff) {
                            if let Some(p) = &song.patterns[m as usize][d as usize] {
                                let cand_floor_val = Self::parse_floor_value(p.floor_name.as_ref());

                                let final_cand_floor = if params.use_official {
                                    if cand_floor_val.is_some()
                                        || Self::diff_group(diff) != params.ref_diff_grp
                                    {
                                        continue;
                                    }
                                    p.level.unwrap_or(0) as f64
                                } else {
                                    match cand_floor_val {
                                        Some(f) => f,
                                        None => continue,
                                    }
                                };

                                if (final_cand_floor - params.ref_floor).abs() > params.floor_range
                                {
                                    continue;
                                }

                                if sid == params.target_song_id
                                    && mode == &params.target_mode
                                    && diff == &params.target_diff
                                {
                                    continue;
                                }

                                candidates.push(RecommendEntry {
                                    song_id: sid,
                                    song_name: song.name.clone(),
                                    composer: song.composer.to_string(),
                                    button_mode: mode.to_string(),
                                    difficulty: diff.to_string(),
                                    level: p.level,
                                    floor: Some(final_cand_floor),
                                    floor_name: p.floor_name.clone(),
                                    rate: None,
                                    is_max_combo: false,
                                });
                            }
                        }
                    }
                }
            }
        }
        candidates
    }

    fn merge_record_rates(&self, candidates: &mut [RecommendEntry]) {
        if !self.rdb.is_ready() {
            return;
        }

        let mut unique_ids = Vec::new();
        for c in candidates.iter() {
            if !unique_ids.contains(&c.song_id) {
                unique_ids.push(c.song_id);
            }
        }

        let rate_map = self.rdb.get_rate_map(&unique_ids);

        for entry in candidates.iter_mut() {
            if let Some(&(rate, is_max_combo)) = rate_map.get(&(
                entry.song_id,
                entry.button_mode.clone(),
                entry.difficulty.clone(),
            )) {
                entry.rate = Some(rate as f64);
                entry.is_max_combo = is_max_combo;
            }
        }
    }

    fn build_floor_cache_index(&self) {
        let mut floor_patterns = HashMap::new();
        let mut record_to_floor_key = HashMap::new();

        for song in &self.vdb.songs {
            let song_id = match song.title.parse::<i32>() {
                Ok(id) => id,
                Err(_) => continue,
            };
            for (m_idx, &mode) in MODES.iter().enumerate() {
                for d_idx in 0..4 {
                    let diff = match d_idx {
                        0 => "NM",
                        1 => "HD",
                        2 => "MX",
                        3 => "SC",
                        _ => unreachable!(),
                    };
                    if let Some(p) = &song.patterns[m_idx][d_idx] {
                        let floor_val;
                        let scale_type;
                        if let Some(f) = Self::parse_floor_value(p.floor_name.as_ref()) {
                            floor_val = f;
                            scale_type = "UNOFFICIAL".to_string();
                        } else {
                            if let Some(level) = p.level {
                                floor_val = level as f64;
                            } else {
                                continue;
                            }
                            scale_type = if SC_GROUP.contains(&diff) {
                                "OFFICIAL_SC".to_string()
                            } else {
                                "OFFICIAL_NHM".to_string()
                            };
                        }

                        let key = FloorCacheKey {
                            button_mode: mode.to_string(),
                            scale_type,
                            floor_millis: Self::floor_to_millis(floor_val),
                        };
                        let record_key = (song_id, mode.to_string(), diff.to_string());
                        floor_patterns
                            .entry(key.clone())
                            .or_insert_with(Vec::new)
                            .push(record_key.clone());
                        record_to_floor_key.insert(record_key, key);
                    }
                }
            }
        }

        let mut cache_guard = overmax_core::lock_or_recover(&self.floor_rate_cache);
        let mut dirty_guard = overmax_core::lock_or_recover(&self.floor_rate_dirty);
        let mut patterns_guard = overmax_core::lock_or_recover(&self.floor_patterns);
        let mut record_to_key_guard = overmax_core::lock_or_recover(&self.record_to_floor_key);

        *patterns_guard = floor_patterns;
        *record_to_key_guard = record_to_floor_key;

        cache_guard.clear();
        dirty_guard.clear();
        for (key, entries) in patterns_guard.iter() {
            cache_guard.insert(key.clone(), FloorRateSummary::new(entries.len()));
            dirty_guard.insert(key.clone(), true);
        }

        self.cache_index_ready.store(true, AtomicOrdering::SeqCst);
    }

    fn ensure_floor_rate_cache(&self) {
        if !self.cache_index_ready.load(AtomicOrdering::SeqCst) {
            self.build_floor_cache_index();
        }

        let (full_dirty, dirty_keys) = self.rdb.consume_dirty_info();
        {
            let mut dirty_guard = overmax_core::lock_or_recover(&self.floor_rate_dirty);
            let patterns_guard = overmax_core::lock_or_recover(&self.floor_patterns);
            let record_to_key_guard = overmax_core::lock_or_recover(&self.record_to_floor_key);

            if full_dirty {
                for key in patterns_guard.keys() {
                    dirty_guard.insert(key.clone(), true);
                }
            } else {
                for record_key in &dirty_keys {
                    if let Some(floor_key) = record_to_key_guard.get(record_key) {
                        dirty_guard.insert(floor_key.clone(), true);
                    }
                }
            }
        }

        let dirty_floor_keys: Vec<FloorCacheKey> = {
            let dirty_guard = overmax_core::lock_or_recover(&self.floor_rate_dirty);
            dirty_guard
                .iter()
                .filter(|(_, &is_dirty)| is_dirty)
                .map(|(k, _)| k.clone())
                .collect()
        };

        if dirty_floor_keys.is_empty() {
            return;
        }

        let mut all_song_ids = Vec::new();
        for song in &self.vdb.songs {
            if let Ok(song_id) = song.title.parse::<i32>() {
                all_song_ids.push(song_id);
            }
        }
        all_song_ids.sort_unstable();
        all_song_ids.dedup();

        let rate_map = self.rdb.get_rate_map(&all_song_ids);

        let mut cache_guard = overmax_core::lock_or_recover(&self.floor_rate_cache);
        let mut dirty_guard = overmax_core::lock_or_recover(&self.floor_rate_dirty);
        let patterns_guard = overmax_core::lock_or_recover(&self.floor_patterns);

        for key in &dirty_floor_keys {
            let entries = patterns_guard.get(key).cloned().unwrap_or_default();
            let mut summary = FloorRateSummary::new(entries.len());
            for record_key in &entries {
                if let Some(&(rate, _)) = rate_map.get(record_key) {
                    if rate > 0.0 {
                        summary.has_record_count += 1;
                        summary.rate_sum += rate as f64;
                    }
                }
            }
            cache_guard.insert(key.clone(), summary);
            dirty_guard.insert(key.clone(), false);
        }
    }

    fn get_summary_from_cache(
        &self,
        button_mode: &str,
        difficulty: &str,
        ref_floor: f64,
        use_official: bool,
        floor_range: f64,
        same_mode_only: bool,
    ) -> FloorRateSummary {
        self.ensure_floor_rate_cache();

        let scale_type = if use_official {
            if SC_GROUP.contains(&difficulty) {
                "OFFICIAL_SC"
            } else {
                "OFFICIAL_NHM"
            }
        } else {
            "UNOFFICIAL"
        };

        let modes = if same_mode_only {
            vec![button_mode]
        } else {
            MODES.to_vec()
        };

        let mut total = 0;
        let mut has_record = 0;
        let mut rate_sum = 0.0;

        let cache_guard = overmax_core::lock_or_recover(&self.floor_rate_cache);
        for (key, summary) in cache_guard.iter() {
            if !modes.contains(&key.button_mode.as_str()) {
                continue;
            }
            if key.scale_type != scale_type {
                continue;
            }

            let key_floor = key.floor_millis as f64 / 1000.0;
            if (key_floor - ref_floor).abs() > floor_range {
                continue;
            }

            total += summary.total_count;
            has_record += summary.has_record_count;
            rate_sum += summary.rate_sum;
        }

        FloorRateSummary {
            total_count: total,
            has_record_count: has_record,
            rate_sum,
        }
    }
}
